llvm/lib/CAS/PluginCAS.cpp

```diff
@@ -6,12 +6,396 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include "PluginAPI.h"
+#include "llvm/ADT/ScopeExit.h"
+#include "llvm/CAS/ActionCache.h"
 #include "llvm/CAS/ObjectStore.h"
+#include "llvm/Support/DynamicLibrary.h"
+#include "llvm/Support/StringSaver.h"
 
 using namespace llvm;
 using namespace llvm::cas;
 
-Expected<std::unique_ptr<ObjectStore>>
-cas::createPluginCAS(StringRef, ArrayRef<std::string>) {
-  report_fatal_error("CAS plugins not implemented");
+namespace {
+
+class PluginCASContext : public CASContext {
+public:
+  void printIDImpl(raw_ostream &OS, const CASID &ID) const final;
+
+  StringRef getHashSchemaIdentifier() const final { return SchemaName; }
+
+  static Expected<std::shared_ptr<PluginCASContext>>
+  create(StringRef PluginPath, StringRef OnDiskPath,
+         ArrayRef<std::pair<StringRef, StringRef>> PluginArgs);
+
+  ~PluginCASContext() {
+    Functions.cas_dispose(c_cas);
+    // Intentionally leak the DLHandle; we have no reason to dlclose it and it
+    // may be unsafe.
+  }
+
+  llcas_functions_t Functions{};
+  llcas_cas_t c_cas = nullptr;
+  std::string SchemaName;
+
+  static Error errorWithConsumed(char *c_err, const llcas_functions_t &Funcs) {
+    Error E = createStringError(inconvertibleErrorCode(), c_err);
+    Funcs.string_dispose(c_err);
+    return E;
+  }
+
+  Error errorWithConsumed(char *c_err) const {
+    return errorWithConsumed(c_err, Functions);
+  }
+};
+
+} // anonymous namespace
+
+void PluginCASContext::printIDImpl(raw_ostream &OS, const CASID &ID) const {
+  ArrayRef<uint8_t> Hash = ID.getHash();
+  char *c_printed_id = nullptr;
+  char *c_err = nullptr;
+  if (Functions.digest_print(c_cas, llcas_digest_t{Hash.data(), Hash.size()},
+                             &c_printed_id, &c_err))
+    report_fatal_error(StringRef(toString(errorWithConsumed(c_err))));
+  OS << c_printed_id;
+  Functions.string_dispose(c_printed_id);
+}
+
+Expected<std::shared_ptr<PluginCASContext>>
+PluginCASContext::create(StringRef PluginPath, StringRef OnDiskPath,
+                         ArrayRef<std::pair<StringRef, StringRef>> PluginArgs) {
+  auto reportError = [PluginPath](const Twine &Description) -> Error {
+    return createStringError(inconvertibleErrorCode(), "error loading '" +
+                                                           PluginPath +
+                                                           "': " + Description);
```

Comment:
I don't want to fight clang-format since it's gating patches; I made this change:
```
    std::error_code EC = inconvertibleErrorCode();
    return createStringError(EC, "error loading '" + PluginPath +
                                     "': " + Description);
```

Reply:
Good enough, thanks!